lighthouse-core/gather/gatherers/start-url.js

```diff
@@ -9,76 +9,90 @@ const Gatherer = require('./gatherer');
-              debugString,
-            };
-          } else {
-            const debugString = `No usable web app manifest found on page ${options.url}`;
-            return {
-              statusCode: -1,
-              debugString,
-            };
-          }
+        const {isReadFailure, reason, startUrl} = this._readManifestStartUrl(manifest);
+        if (isReadFailure) {
+          return {statusCode: -1, debugString: reason};
         }
 
-        if (manifest.value.start_url.debugString) {
-          // Even if the start URL had an error, the browser will still supply a fallback URL.
-          // Therefore, we only set the debugString here and continue with the fetch.
-          return {
-            statusCode: -1,
-            debugString: manifest.value.start_url.debugString,
-          };
-        }
+        return this._attemptManifestFetch(options, startUrl);
+      }).catch(() => {
+        return {statusCode: -1, debugString: 'Unable to fetch start URL via service worker'};
+      });
+  }
+
+  /**
+   * Read the parsed manifest and return failure reasons or the startUrl
+   * @param {Manifest} manifest
+   * @return {{isReadFailure: !boolean, reason: ?string, startUrl: ?startUrl}}
+   */
+  _readManifestStartUrl(manifest) {
+    if (!manifest || !manifest.value) {
+      const detailedMsg = manifest && manifest.debugString;
 
-        const startUrl = manifest.value.start_url.value;
+      if (detailedMsg) {
+        return {isReadFailure: true, reason: `Error fetching web app manifest: ${detailedMsg}`};
+      } else {
+        return {isReadFailure: true, reason: `No usable web app manifest found on page`};
+      }
+    }
 
-        return (new Promise(resolve => {
-          options.driver.on('Network.responseReceived', function responseReceived({response}) {
-            if (response.url === startUrl) {
-              options.driver.off('Network.responseReceived', responseReceived);
+    // Even if the start URL had an error, the browser will still supply a fallback URL.
+    // Therefore, we only set the debugString here and continue with the fetch.
+    if (manifest.value.start_url.debugString) {
+      return {isReadFailure: true, reason: manifest.value.start_url.debugString};
+    }
 
-              if (!response.fromServiceWorker) {
-                return resolve({
-                  statusCode: -1,
-                  debugString: 'Unable to fetch start URL via service worker',
-                });
-              }
+    return {isReadFailure: false, startUrl: manifest.value.start_url.value};
+  }
+
+  /**
+   * Try to `fetch(start_url)`, return true if fetched by SW
+   * @param {*} options
+   * @param {!string} startUrl
+   * @return {Promise<{statusCode: ?number, debugString: ?string}>}
+   */
+  _attemptManifestFetch(options, startUrl) {
+    // Wait up to 3s to get a matched network request from the fetch() to work
+    const timeoutPromise = new Promise(resolve => setTimeout(_ =>
```

Comment:
do we want to hold off on adding this and just have a general "add generalized timeout method to driver commands" issue? Only bring this up because we've never had reports of this timing out, so not sure if it's worth adding in various implementations of timeouts spread out over a few different gatherers

Reply:
i'm ok with delaying.

up till now the starturl gatherer uses `fetch('${url}').then(response => response.status)`. That's pretty direct and I wouldn't expect it to hang forever.

Now we're fetching and observing all network requests until we see one with a matching url. it's different and i'm not 100% we'll always have a match. redirects _shouldn't_ time out but ... it probably will mess up.